vdr/didweb/manager.go

```diff
@@ -107,64 +113,62 @@ func (m Manager) AddVerificationMethod(_ context.Context, _ did.DID, _ managemen
 	for _, opt := range opts.All() {
-		if !newDID.Empty() {
-			return nil, nil, errors.New("multiple DID options provided")
-		}
-		switch option := opt.(type) {
-		case tenantOption:
-			newDID = m.rootDID
-			newDID.ID += ":" + m.tenantPath + ":" + option.id
-		case didOption:
-			newDID = option.value
+		switch opt.(type) {
+		case rootDIDOption:
+			newDID = &m.rootDID
 		default:
-			return nil, nil, fmt.Errorf("unknown option: %T", option)
+			return nil, nil, fmt.Errorf("unknown option: %T", opt)
 		}
 	}
-	if newDID.Empty() {
-		newDID = m.rootDID
-		newDID.ID += ":" + m.tenantPath + ":" + uuid.NewString()
+	if newDID == nil {
+		newDID, err = did.ParseDID(fmt.Sprintf("%s:iam:%s", m.rootDID.String(), uuid.NewString()))
 	}
-	parsedNewDID, err := did.ParseDID(newDID.String()) // make sure internal state is consistent (DecodedID)
 	if err != nil {
-		return nil, nil, fmt.Errorf("invalid new DID: %s: %w", newDID.String(), err)
+		return nil, nil, fmt.Errorf("parse new DID: %w", err)
 	}
-	return m.create(ctx, *parsedNewDID)
-}
+	tx := m.db.Begin()
+	defer tx.Rollback()
 
-func (m Manager) create(ctx context.Context, newDID did.DID) (*did.Document, crypto.Key, error) {
-	// in any case, the DID to created (with or without path) should be scoped to the configured URL (translated to DID)
-	if !strings.HasPrefix(newDID.String(), m.rootDID.String()) {
-		return nil, nil, fmt.Errorf("invalid DID, does not match configured base URL, translated to DID: %s", m.rootDID.String())
+	documentStore := sql.NewDIDDocumentManager(tx)
+
+	_, err = documentStore.Latest(*newDID)
+	if err == nil {
+		return nil, nil, management.ErrDIDAlreadyExists
 	}
-	// if it contains an optional path, it must be in format did:web:<host>:iam:<tenant>
-	idParts := strings.Split(newDID.ID, ":")
-	if (len(idParts) > 2 && (idParts[1] != m.tenantPath || len(idParts) > 3)) ||
-		// it must not contain empty path parts
-		strings.HasSuffix(newDID.ID, ":") || strings.HasSuffix(newDID.ID, "::") {
-		return nil, nil, errors.New("invalid path in did:web DID, it must follow the pattern 'did:web:<host>:" + m.tenantPath + ":<tenant>'")
+	if !errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, nil, err
 	}
 
-	// Check if it doesn't already exist. Otherwise, it fail later on (unique key constraint) but we might end up with an orphaned private key.
-	exists, err := m.IsOwner(ctx, newDID)
+	verificationMethodKey, verificationMethod, err := m.createVerificationMethod(ctx, *newDID)
 	if err != nil {
 		return nil, nil, err
 	}
-	if exists {
-		return nil, nil, management.ErrDIDAlreadyExists
-	}
-
-	verificationMethodKey, verificationMethod, err := m.createVerificationMethod(ctx, newDID)
+	vmAsJson, err := json.Marshal(verificationMethod)
 	if err != nil {
 		return nil, nil, err
 	}
-	if err := m.store.create(newDID, *verificationMethod); err != nil {
-		return nil, nil, fmt.Errorf("store new DID: %w", err)
+
+	sqlDid := sql.DID{
+		ID:      newDID.String(),
+		Subject: newDID.String(),
```

Comment:
shouldn't this be a UUID (subject) and be supplied through the options?

Reply:
correct, but allowing a subject to be passed through the API would be part of another PR. Added a todo as comment